helix-term/src/ui/completion.rs

```diff
@@ -213,14 +227,27 @@ impl Completion {
 
             let (view, doc) = current!(editor);
 
-            // if more text was entered, remove it
-            doc.restore(view, &savepoint);
-
             match event {
-                PromptEvent::Abort => {
-                    editor.last_completion = None;
-                }
+                PromptEvent::Abort => {}
                 PromptEvent::Update => {
+                    // Update creates "ghost" transactiosn which are not send to the
+                    // lsp server to avoid messing up rerequesting completions. Once a
+                    // completion has been selected (with) tab it's always accepted whenever anything
+                    // is typed. The only way to avoid that is to explicitly abort the completion
+                    // with esc/c-c. This will remove the "ghost" transaction.
+                    //
+                    // The ghost transaction is modeled with a transaction that is not send to the LS.
+                    // (apply_temporary) and a savepoint. It's extremly important this savepoint is restored
+                    // (also without sending the transaction to the LS) *before any further transaction is applied*.
+                    // Otherwise incremental sync breaks (since the state of the LS doesn't match the state the transaction
+                    // is applied to).
```

Comment:
```suggestion
                    // Update creates "ghost" transactions which are not send to the
                    // lsp server to avoid messing up re-requesting completions. Once a
                    // completion has been selected (with) tab it's always accepted whenever anything
                    // is typed. The only way to avoid that is to explicitly abort the completion
                    // with esc/c-c. This will remove the "ghost" transaction.
                    //
                    // The ghost transaction is modeled with a transaction that is not sent to the LS.
                    // (apply_temporary) and a savepoint. It's extremely important this savepoint is restored
                    // (also without sending the transaction to the LS) *before any further transaction is applied*.
                    // Otherwise incremental sync breaks (since the state of the LS doesn't match the state the transaction
                    // is applied to).
```

Reply:
minor nit about the "selected with tab" part - I think that could be clearer as "selected (for example with tab or C-n/C-p) since there are a few ways of selecting an item. It's possible that tab could get more complicated behavior in the future too if we use it to jump between tabstops so this might not age well

Reply:
yeah that's a good point the tab was always supposed to be in brackets and was means as an example just to make sure it's clear what selected meant. I updated the comment to be more clear